### ISSUE TITLE
Direct saving of key-values to database

### ISSUE THREAD

**Issue by joozek78** [Starcounter/Starcounter#3911]
Regular way to save somethings into database is by using [Database] Attribute on a class and then use it:
```c#
[Database]
public class TheClass
{
  public int TheField;
}

// ...
var obj = new TheClass() { TheField = 42 }; // saves to database magically
```

Now I want to store objects whose schemas are unknown to me in compile time. I'd like to do something like this:
```c#
var obj = Db.Store(new Dictionary<string, int>() {{"TheField", 3}}); // this would save the object equivalent to one from the previous example
```

Of course, I could just create a [Database] class with Dictionary in it, but this feels like working around something already implemented (we store arbitrary documents, right?). Can I achieve something like I described?

**Comment by mtornwall** [comment]
> Of course, I could just create a [Database] class with Dictionary in it

Sorry, `Dictionary` is not a supported Starcounter type.

> Can I achieve something like I described?

To my knowledge there is no way of accomplishing what you are describing using the current version of Starcounter.

**Comment by joozek78** [comment]
> To my knowledge there is no way of accomplishing what you are describing using the current version of Starcounter.

Would it be possible in future versions?

Does weaver do what I just described? I.e. translate [Database] classes to the low-level (key-value) API?

**Comment by mtornwall** [comment]
> Would it be possible in future versions?

I've been told that some potentially related work has been done in order to support dynamically typed host languages. Perhaps @eriohl can shed some light on the matter?

> Does weaver do what I just described? I.e. translate [Database] classes to the low-level (key-value) API?

Not to my knowledge.

**Comment by mtornwall** [comment]
Sorry, I'm tired - I thought you were asking if you can use Starcounter as a key-value store.

Now I see what you mean, but I'm not the right person to answer.

I'm sure @per-samuelsson can tell you about the weaver.

**Comment by k-rus** [comment]
I don't really understand requirements in OP. The second code snippet looks like as just key-value store. 

However, OP also says about unknown schema. How are you going to use C# for this? The proposed solution with dictionary looks just as a hack and you can do the same with Starcounter by using it as key-value store by having single table and index on the key. I see the only possibility is to use dynamic properties on dynamic types.

**Comment by joozek78** [comment]
I wouldn't know the classes in compile time (my understanding tells that's when weaver does some magic to access low-level API, correct me if I'm wrong), but I would in runtime. So what I want to do is (to some degree) like invoking weaver during runtime on a generated class.

**Comment by k-rus** [comment]
> I wouldn't know the classes in compile time (my understanding tells that's when weaver does some magic to access low-level API, correct me if I'm wrong), but I would in runtime. So what I want to do is (to some degree) like invoking weaver during runtime on a generated class.

I'm sorry, but I still don't understand. You are mixing your need with a possible solution, so  I cannot really get what you want.

How can you operate on an object in C# if you don't know the type at compile time? I believe we are in C# code host in this issue.

**Comment by joozek78** [comment]
tl;dr I want to be able to generate a new class in runtime and save an instance to database.

Well, not exactly that, but it's equivalent :)
[This article](http://starcounter.io/guides/tools/weaver/) shows weaver using the function `Db.ReadString`. I'd like to use this function (and its siblings) directly

**Comment by k-rus** [comment]
> tl;dr I want to be able to generate a new class in runtime and save an instance to database.

I don't understand how you will do it in C# in general. Do you have any nice suggestion which C# syntax you will utilise for this?

We made some steps in the direction of dynamic types. Still work needs to be done.

I don't think it is good idea to provide low level API to App developers, which will be very low, since you cannot query without loaded database classes.

**Comment by per-samuelsson** [comment]
> This article shows weaver using the function Db.ReadString. I'd like to use this function (and its siblings) directly.

The code there is just pseudo-code. But with similar semantics to what is emitted by the weaver.

If I understand you correct, you want support for a schema-less solution where you could at any time define an arbitrary type, then instantiate it, then CRUD those instances. We don't have such an API today unfortunately.

**Comment by per-samuelsson** [comment]
This is just me thinking out loud, but we have had discussions about this and a similar case we don't support today either - that of allowing dynamically loaded code to be weaved on the fly, and made usable.

Today, our bootstrapper will analyze your application at the time it is booted in the host. It will synchronize its host level schema (i.e. the database class graph) with the current databases logical schema, and then any CRUD operations you make on that class get reflected in the database.

If you would do this:

```cs
var assembly = Assembly.LoadFrom("some/random/path/with/app.dll");
```

and `app.dll` would contain database classes, those classes would not be usable in the same maner as those loaded per the description above. In a future host implementation, we want to support that scenario too. And if we do, I think it make sense to support what you are asking for also (it's basically a similar approach).

**Comment by joozek78** [comment]
Thank you for answering this.

>I don't understand how you will do it in C# in general. Do you have any nice suggestion which C# syntax you will utilise for this?

Just for reference, you could directly emit IL that defines a class. [This SO question](http://stackoverflow.com/questions/3862226/how-to-create-dynamically-a-class-in-c)  seems to point in proper direction. For the record, I haven't (yet) tried it.

**Comment by per-samuelsson** [comment]
> Just for reference, you could directly emit IL that defines a class.

Exactly. Using `Reflection.Emit`, or Roslyn, and load the generated code would be use cases similar to that of dynamically loading a pre-compiled assembly as I described above. It's the same thing really, just that `Reflection.Emit` and Roslyn would compile the type on the fly, while *loading* it is what matter to us. As soon as we support that case, whatever technology you use to generate code is transparent to us as an engine.

**Comment by mtornwall** [comment]
> Just for reference, you could directly emit IL that defines a class. 

You could, but this is probably one of the cases where it's easier to emit C# code and invoke the compiler at runtime.

**Comment by joozek78** [comment]
Just for reference (again :D)
I don't want to really generate the class, I just want to access the database APIs on lower level (directly reading a field of an object)

**Comment by k-rus** [comment]
> Just for reference, you could directly emit IL that defines a class.

This allows to create the class and objects. But how to write logic, which will work with objects. Providing completely new API is not a good solution. 



> I don't want to really generate the class, I just want to access the database APIs on lower level (directly reading a field of an object)

IMHO, this approach will bring us to troubles. It will be nice to use something native, instead of providing some new API, which will reduce developers productivity and will bind us from new changes.

**Comment by joozek78** [comment]
>> I don't want to really generate the class, I just want to access the database APIs on lower level (directly reading a field of an object)
> IMHO, this approach will bring us to troubles. It will be nice to use something native, instead of providing some new API, which will reduce developers productivity and will bind us from new changes.

What do you mean by "native"? And by "new"? Isn't this API already there, used by the weaver? What I want would not be used by apps developers to write an app, but rather to create a productivity-boosting library (like I'm creating)

**Comment by k-rus** [comment]
> What do you mean by "native"?

Native for C# developers



> Isn't this API already there, used by the weaver?

Current weaver only uses CRUD API, which requires much more than IL instructions. Furthermore, it does not cover querying at all.



> What I want would not be used by apps developers to write an app, but rather to create a productivity-boosting library (like I'm creating)

I see. So it is some internal general tool to be used by App developers. This explains why you have this need.

For the level of meta-programming it should be possible to provide you necessary facilities as @per-samuelsson described.
